internal/retry/buffer.go

```diff
@@ -0,0 +1,31 @@
+package retry
+
+// FillBufferFunc callbacks are used by FillBuffer to populate
+// the contents of the passed buffer.
+type FillBufferFunc func([]byte) error
+
+// FillBuffer will fill the contents of a buffer by calling the fill callback.
+// The function will be retried if the fill function returns the retryOn error.
+// The buffer starts off at startSize and will grow up to and including maxSize.
+// Any other error returned by the fill function will cancel retries and
+// the error will be returned.
+func FillBuffer(
+	fill FillBufferFunc, retryOn error, startSize, maxSize int) ([]byte, error) {
```

Comment:
I like the idea, but maybe this can be simplified a little more? If possible, I'd like to see the retryOn error dropped. Instead, the FillBufferFunc could return (bool, error), and on 'false', the retry can be done. Most (not sure all?) errors that relate to allocating some C-memory have defined errors, but maybe there can be multiple errors, or some other undefined condition. Having that spelled out in the FillBufferFunc probably makes the code easier to understand.

Reply:
I also think returning a bool in retry/loop/walk sub-functions is a relatively common Go approach?